lib/openhab/dsl/debouncer.rb

```diff
@@ -0,0 +1,241 @@
+      #   The debouncer will continue to hold until the interval between two calls is longer
+      #   than the idle time or until the maximum interval between executions is reached.
+      #
+      # @return [void]
+      #
+      def initialize(for:, leading: false, idle_time: nil)
+        @interval = binding.local_variable_get(:for)
+        return unless @interval
+
+        @interval = (@interval..) unless @interval.is_a?(Range)
+
+        @leading = leading
+        @idle_time = idle_time
+        @mutex = Mutex.new
+        @block = nil
+        @debounce_timer = nil
+        reset
+      end
+
+      #
+      # Debounce calls to the given block.
+      #
+      # This method is meant to be called repeatedly with the same given block.
+      # However, if no block is given, it will call and debounce the previously given block
+      #
+      # @yield Block to be debounced
+      #
+      # @return [void]
+      #
+      # @example Basic trailing edge debouncing
+      #   debouncer = Debouncer.new(for: 1.minute)
+      #   (1..100).each do
+      #     debouncer.call { logger.info "I won't log more often than once a minute" }
+      #     sleep 20 # call the debouncer every 20 seconds
+      #   end
+      #
+      # @example Call the previous debounced block
+      #   debouncer = Debouncer.new(for: 1.minute)
+      #   debouncer.call { logger.info "Hello. It is #{Time.now}" } # First call to debounce
+      #
+      #   after(20.seconds) do |timer|
+      #     debouncer.call # Call the original block above
+      #     timer.reschedule unless timer.cancelled?
+      #   end
+      #
+      def call(&block)
+        @block = block if block
+        raise ArgumentError, "No block has been provided" unless @block
+
+        return call! unless @interval # passthrough mode, no debouncing when @interval is nil
+
+        now = ZonedDateTime.now
+
+        if leading?
+          leading_edge_debounce(now)
+        else
+          trailing_edge_debounce(now)
+        end
+        @mutex.synchronize { @last_timestamp = now }
+      end
+
+      #
+      # Execute the latest block passed to the {#debounce} call regardless of any debounce settings.
+      #
+      # @return [Object] The return value of the block
+      #
+      def call!
+        @block.call
+      end
+
+      #
+      # Resets the debounce period and cancels any outstanding block executions of a trailing edge debouncer.
+      #
+      # - A leading edge debouncer will execute its block on the next call and start a new debounce period.
+      # - A trailing edge debouncer will reset its debounce timer and the next call will become the start
+      #   of a new debounce period.
+      #
+      # @return [Boolean] True if a pending execution was cancelled.
+      #
+      def reset
+        @mutex.synchronize do
+          if leading?
+            @last_timestamp = @leading_timestamp = ZonedDateTime.of(1900, 1, 1, 0, 0, 0, 0,
```

Comment:
`= @interval.begin.ago`? then you don't need a magic timestamp, and it should belong enough ago (or possibly subtract one more second from it)

Reply:
Because in some specs I did Timecop.freeze(Time.at(0)) I need a date/time at epoch 0 or earlier. My reason for choosing this form is because I think it would require the least computation, even compared to using @interval.begin.ago (which would involve calling now + minus). The other alternative is to use a constant here and do the calculation elsewhere.

I see `reset` as something that's very rarely called anyway.

If you still think I should change it, including changing the specs, I'll be happy to do it if I understand the reason.

Reply:
The specs shouldn't need to be changed - that's the beauty of it. Even if for some reason we're at `Time.now(-1)`, using `.ago` will put us even further back in time.